embassy-nrf/src/pwm.rs

```diff
@@ -1,16 +1,305 @@
 use embassy::util::Unborrow;
 use embassy_hal_common::unborrow;
 
 use crate::gpio::sealed::Pin as _;
-use crate::gpio::OptionalPin as GpioOptionalPin;
+use crate::gpio::{AnyPin, OptionalPin as GpioOptionalPin};
 use crate::interrupt::Interrupt;
 use crate::pac;
+use crate::util::slice_in_ram_or;
+
+/// Interface to the PWM peripheral
+pub struct SimplePwm<'d, T: Instance> {
+    phantom: PhantomData<&'d mut T>,
+    duty: [u16; 4],
+    ch0: Option<AnyPin>,
+    ch1: Option<AnyPin>,
+    ch2: Option<AnyPin>,
+    ch3: Option<AnyPin>,
+}
+
+pub struct SequencePwm<'d, T: Instance> {
+    phantom: PhantomData<&'d mut T>,
+    ch0: Option<AnyPin>,
+    ch1: Option<AnyPin>,
+    ch2: Option<AnyPin>,
+    ch3: Option<AnyPin>,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[non_exhaustive]
+pub enum Error {
+    /// Max Sequence size is 32767
+    SequenceTooLong,
+    /// Min Sequence count is 1
+    SequenceTimesAtLeastOne,
+    /// EasyDMA can only read from data memory, read only buffers in flash will fail.
+    DMABufferNotInDataMemory,
+}
+
+impl<'d, T: Instance> SequencePwm<'d, T> {
+    /// Creates the interface to a PWM Sequence interface.
+    ///
+    /// Must be started by calling `start`
+    ///
+    /// # Safety
+    ///
+    /// The returned API is safe unless you use `mem::forget` (or similar safe
+    /// mechanisms) on stack allocated buffers which which have been passed to
+    /// [`new()`](SequencePwm::new).
+    #[allow(unused_unsafe)]
+    pub fn new<'a>(
+        _pwm: impl Unborrow<Target = T> + 'd,
+        ch0: impl Unborrow<Target = impl GpioOptionalPin> + 'd,
+        ch1: impl Unborrow<Target = impl GpioOptionalPin> + 'd,
+        ch2: impl Unborrow<Target = impl GpioOptionalPin> + 'd,
+        ch3: impl Unborrow<Target = impl GpioOptionalPin> + 'd,
+        config: SequenceConfig,
+        sequence: &'a [u16],
+    ) -> Result<Self, Error> {
+        slice_in_ram_or(sequence, Error::DMABufferNotInDataMemory)?;
+
+        if sequence.len() > 32767 {
+            return Err(Error::SequenceTooLong);
+        }
+
+        unborrow!(ch0, ch1, ch2, ch3);
+
+        let r = T::regs();
+
+        if let Some(pin) = ch0.pin_mut() {
+            pin.set_low();
+            pin.conf().write(|w| w.dir().output());
+        }
+        if let Some(pin) = ch1.pin_mut() {
+            pin.set_low();
+            pin.conf().write(|w| w.dir().output());
+        }
+        if let Some(pin) = ch2.pin_mut() {
+            pin.set_low();
+            pin.conf().write(|w| w.dir().output());
+        }
+        if let Some(pin) = ch3.pin_mut() {
+            pin.set_low();
+            pin.conf().write(|w| w.dir().output());
+        }
+
+        // if NoPin provided writes disconnected (top bit 1) 0x80000000 else
+        // writes pin number ex 13 (0x0D) which is connected (top bit 0)
+        r.psel.out[0].write(|w| unsafe { w.bits(ch0.psel_bits()) });
```

Comment:
do we have to 'disconnect' pins on drop as well?

Reply:
not needed, writing enable=0 already makes PSELs not apply

Reply:
Or maybe not? nrfx does disconnect them
https://github.com/NordicSemiconductor/nrfx/blob/master/drivers/src/nrfx_uarte.c#L341
https://github.com/NordicSemiconductor/nrfx/blob/master/drivers/src/nrfx_pwm.c#L207

I'm not sure if it's a "just in case" thing or the hardware actually requires it...

Reply:
I threw it in